File: merge-sort/merge-sort.js
var mergeSort = function (arr) {
  var left, right, middle;

  if (arr.length === 0 || arr.length === 1) {
    return arr;
  }

  middle = parseInt(arr.length / 2);
  left = arr.slice(0, middle);
  right = arr.slice(middle, arr.length);

  return merge(mergeSort(left), mergeSort(right));
};


function merge (a, b) {
  var result, i, j;

  result = [];
  i = 0;
  j = 0;

  while (i < a.length && j < b.length) {
    if (a[i] < b[j]) {
      result.push(a[i]);
      i += 1;
    } else {
      result.push(b[j]);
      j += 1;
    }
  }

  return result.concat(a.slice(i, a.lenght), b.slice(j, b.length));
}

File: merge-sort/merge-sort.rb
def merge_sort(arr)
  return arr if arr.count == 0 || arr.count == 1

  middle = arr.length / 2
  left = arr[0...middle]
  right = arr[middle..-1]

  merge(merge_sort(left), merge_sort(right))
end

def merge(a, b)
  result = []

  i, j = 0, 0;

  while i < a.count && j < b.count
    if a[i] < b[j]
      result << a[i]
      i += 1
    else
      result << b[j]
      j += 1
    end
  end

  result.concat(a[i..-1]).concat(b[j..-1])
end
